src/creature_tracker.cpp

```diff
@@ -69,6 +85,20 @@ bool Creature_tracker::add( monster &critter )
         return false;
     }
 
+    if( critter.get_id() == 0 ) {
+        critter.set_id( g->assign_creature_id() );
+    }
+
+    const Creature *other = g->critter_with_id( critter.get_id() );
+    while( other != nullptr && other != &critter ) {
+        debugmsg( "%s had id %ld, which belongs to %s",
+                  critter.disp_name().c_str(), critter.get_id(),
+                  other->disp_name().c_str() );
+        critter.set_id( g->assign_creature_id() );
```

Comment:
I'd go a step further and say this simply shouldn't be checked, we either have a unique id or we don't, if there's a collision this isn't guaranteed to catch it anyway.